src/interpreter.rs

```diff
@@ -23,57 +16,38 @@ fn is_truthy(val: &LoxValue) -> bool {
     }
 }
 
-impl PartialEq for LoxValue {
-    fn eq(&self, other: &Self) -> bool {
-        println!("{} {}", self, other);
-        match (self, other) {
-            (LoxValue::Nil, LoxValue::Nil) => false,
-            (LoxValue::Boolean(a), LoxValue::Boolean(b)) => a == b,
-            (LoxValue::Number(a), LoxValue::Number(b)) => a == b,
-            (LoxValue::String(a), LoxValue::String(b)) => a == b,
-            (_, _) => false,
-        }
-    }
-}
-
-impl TryFrom<&Token> for LoxValue {
-    type Error = InterpreterError;
-
-    fn try_from(value: &Token) -> Result<Self, Self::Error> {
-        match &value.t {
-            TokenType::Nil => Ok(Self::Nil),
-            TokenType::True => Ok(Self::Boolean(true)),
-            TokenType::False => Ok(Self::Boolean(false)),
-            TokenType::Number(v) => Ok(Self::Number(*v)),
-            TokenType::String(v) => Ok(Self::String(v.to_string())),
-            _ => Err(InterpreterError::from_token(value, "Unexpected value")),
-        }
-    }
-}
-
-impl fmt::Display for LoxValue {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match self {
-            LoxValue::Number(v) => write!(f, "{}", v),
-            LoxValue::String(v) => write!(f, "{}", v),
-            LoxValue::Boolean(v) => write!(f, "{}", v),
-            LoxValue::Nil => write!(f, "nil"),
-        }
-    }
-}
-
 pub struct InterpreterError {
     line: usize,
     location: String,
     message: String,
+    pub return_value: Option<LoxValue>,
 }
 
 impl InterpreterError {
+    pub fn new(line: usize, location: impl AsRef<str>, message: impl AsRef<str>) -> Self {
+        Self {
+            line,
+            location: location.as_ref().to_string(),
+            message: message.as_ref().to_string(),
+            return_value: None,
+        }
+    }
+
```

Comment:
I like the generic `impl AsRef<str>` here! I'm not too familiar with `impl Trait`.. Is that equivalent to doing `pub fn new<T: AsRef<str>>(...)`? I sort of recall that it's not fully monomorphic and can only be used with one type, but I could be wrong.

Reply:
I [looked it up](https://stackoverflow.com/questions/47514930/what-are-the-differences-between-an-impl-trait-argument-and-generic-function-par) and they appear to be equivalent except for being able to specify the type.

```rust
fn foo<T: Trait>(t: T)
fn bar(t: impl Trait)

foo::<u32>(0) // this is allowed
bar::<u32>(0) // this is not
```